internal/internal.go

```diff
@@ -223,25 +236,49 @@ func (p *envoyExtAuthzGrpcServer) Check(ctx context.Context, req *ext_authz.Chec
 		return nil, err
 	}
 
-	bs, err := json.Marshal(req)
-	if err != nil {
-		return nil, err
+	var req *ext_authz_v3.CheckRequest
+	var bs []byte
+	var path, body string
+	var headers map[string]string
+
+	// NOTE: The path/body/headers blocks look silly, but they allow us to retrieve
+	//       the parts of the incoming request we care about, without having to convert
+	//       the entire v2 message into v3. It's nested, each level has a different type,
+	//       etc -- we only care for its JSON representation as fed into evaluation later.
+	switch r := r.(type) {
+	case *ext_authz_v3.CheckRequest:
+		req = r
+		bs, err = protojson.Marshal(req) // TODO(sr): Note that the encoding might have changed, figure out how.
```

Comment:
[The linked doc](https://developers.google.com/protocol-buffers/docs/proto3#json) explains how pb is supposed to be turned into json, but it doesn't show the differences from encoding/json... For example encoding/json won't match the spec'ed way to marshal a timestamp value:

```interactive
$ cat main.go
package main

import (
        "fmt"
        "time"
        "encoding/json"

        "google.golang.org/protobuf/types/known/timestamppb"
        "google.golang.org/protobuf/encoding/protojson"
)

func main() {
        ts := timestamppb.New(time.Now())

        bs0, err := json.Marshal(ts)
        if err != nil {
                panic(err)
        }

        bs1, err := protojson.Marshal(ts)
        if err != nil {
                panic(err)
        }

        fmt.Printf("encoding/json: %q\n", bs0)
        fmt.Printf("protojson: %q\n", bs1)
}
$ go run main.go
encoding/json: "{\"seconds\":1605811323,\"nanos\":912400000}"
protojson: "\"2020-11-19T18:42:03.912400Z\""
$
```

It's still logically possible that it doesn't matter how we marshal a CheckRequest (v2, v3), but it would be better to have a close look at all the values that can appear in attribute context. I'll take care of that as I compare v2 and v3 CheckRequests.

Either way, for v3, I think we should go with protojson, as it fulfils the protobuf specification. The question above is a) what this means concretely, and b) how big an impact it is for people writing policies against the JSON representation of a CheckRequest. (The intro of v3 is a good moment to fix this, since no user will switch accidentally from v2 to v3, but only on purpose.)